# Lab | Data Cleaning and Formatting

# Challenge 1: Data Cleaning and Formatting

## Exercise 1: Cleaning Column Names

In [2]:
import pandas as pd

#taking a first look at the dataframe
url = "https://raw.githubusercontent.com/data-bootcamp-v4/data/main/file1.csv"
df = pd.read_csv(url)
print(df.shape)
display(df.head())

#identifying any column names that need to be modified
cols = df.columns
column_names = pd.DataFrame(columns=cols)
print('Current column names:')
display(column_names.T)

new_cols = []
for col in cols:
    #Column names should be in lower case
    new_cols.append(col.lower()) 
    
#White spaces in column names should be replaced by _
new_cols = [col.replace(' ', '_') for col in new_cols] 
df.columns = new_cols
new_cols = pd.DataFrame(columns = new_cols) 
print('Modified column names:')
display(new_cols.T)

#st could be replaced for state
df = df.rename(columns = {'st':'state'})

display(df.head())
print('Data types:')
display(df.dtypes)

(4008, 11)


Customer          ST GENDER             Education Customer Lifetime Value  \
0  RB50392  Washington    NaN                Master                     NaN   
1  QZ44356     Arizona      F              Bachelor              697953.59%   
2  AI49188      Nevada      F              Bachelor             1288743.17%   
3  WW63253  California      M              Bachelor              764586.18%   
4  GA49547  Washington      M  High School or Below              536307.65%   

    Income  Monthly Premium Auto Number of Open Complaints     Policy Type  \
0      0.0                1000.0                    1/0/00   Personal Auto   
1      0.0                  94.0                    1/0/00   Personal Auto   
2  48767.0                 108.0                    1/0/00   Personal Auto   
3      0.0                 106.0                    1/0/00  Corporate Auto   
4  36357.0                  68.0                    1/0/00   Personal Auto   

   Vehicle Class  Total Claim Amount  
0  Four-Door Car            2.704934  
1  Four-Door Car         1131.464935  
2   Two-Door Car          566.472247  
3            SUV          529.881344  
4  Four-Door Car           17.269323

Current column names:


Empty DataFrame
Columns: []
Index: [Customer, ST, GENDER, Education, Customer Lifetime Value, Income, Monthly Premium Auto, Number of Open Complaints, Policy Type, Vehicle Class, Total Claim Amount]

Modified column names:


Empty DataFrame
Columns: []
Index: [customer, st, gender, education, customer_lifetime_value, income, monthly_premium_auto, number_of_open_complaints, policy_type, vehicle_class, total_claim_amount]

customer       state gender             education customer_lifetime_value  \
0  RB50392  Washington    NaN                Master                     NaN   
1  QZ44356     Arizona      F              Bachelor              697953.59%   
2  AI49188      Nevada      F              Bachelor             1288743.17%   
3  WW63253  California      M              Bachelor              764586.18%   
4  GA49547  Washington      M  High School or Below              536307.65%   

    income  monthly_premium_auto number_of_open_complaints     policy_type  \
0      0.0                1000.0                    1/0/00   Personal Auto   
1      0.0                  94.0                    1/0/00   Personal Auto   
2  48767.0                 108.0                    1/0/00   Personal Auto   
3      0.0                 106.0                    1/0/00  Corporate Auto   
4  36357.0                  68.0                    1/0/00   Personal Auto   

   vehicle_class  total_claim_amount  
0  Four-Door Car            2.704934  
1  Four-Door Car         1131.464935  
2   Two-Door Car          566.472247  
3            SUV          529.881344  
4  Four-Door Car           17.269323

Data types:


customer                      object
state                         object
gender                        object
education                     object
customer_lifetime_value       object
income                       float64
monthly_premium_auto         float64
number_of_open_complaints     object
policy_type                   object
vehicle_class                 object
total_claim_amount           float64
dtype: object

## Exercise 2: Cleaning invalid Values

In [3]:
#display unique values
cols_to_replace = ['gender', 'state', 'education', 'customer_lifetime_value', 
                   'vehicle_class']
for column in cols_to_replace:
    unique_values = df[column].unique()
    print(f"\nUnique values for {column}:\n{unique_values}")

#standartize values
df['gender'] = df['gender'].replace({'Femal':'F', 'female':'F', 'Male':'M'})
df['state'] = df['state'].replace({'AZ':'Arizona', 'Cali':'California', 
                                  'WA':'Washington'})  
df['education'] = df['education'].replace({'Bachelors':'Bachelor'})  
df['customer_lifetime_value'] =df['customer_lifetime_value'].str.replace('%','') 
df['vehicle_class'] =df['vehicle_class'].replace({'Sports Car':'Luxury', 
                                                  'Luxury SUV':'Luxury', 
                                                  'Luxury Car':'Luxury'}) 



Unique values for gender:
[nan 'F' 'M' 'Femal' 'Male' 'female']

Unique values for state:
['Washington' 'Arizona' 'Nevada' 'California' 'Oregon' 'Cali' 'AZ' 'WA'
 nan]

Unique values for education:
['Master' 'Bachelor' 'High School or Below' 'College' 'Bachelors' 'Doctor'
 nan]

Unique values for customer_lifetime_value:
[nan '697953.59%' '1288743.17%' ... '2031499.76%' '323912.47%'
 '899704.02%']

Unique values for vehicle_class:
['Four-Door Car' 'Two-Door Car' 'SUV' 'Luxury SUV' 'Sports Car'
 'Luxury Car' nan]


## Exercise 3: Formatting data types

In [4]:
#transform to numeric
df['customer_lifetime_value'] = pd.to_numeric(df['customer_lifetime_value'], errors = 'coerce')
display(df.dtypes['customer_lifetime_value'])
df['number_of_open_complaints'].unique()

#fix the format
df['number_of_open_complaints'] = df['number_of_open_complaints'].apply(lambda x: int(str(x).split('/')[1]) if '/' in str(x) else x)
df['number_of_open_complaints'].unique()



dtype('float64')

array([ 0.,  2.,  1.,  3.,  5.,  4., nan])

## Exercise 4: Dealing with Null values

In [5]:
# Identify columns with null values and count the null values in each column
null_counts = df.isnull().sum()

# Display columns with null values and their respective counts
print("Columns with Null Values:")
print(null_counts)

Columns with Null Values:
customer                     2937
state                        2937
gender                       3054
education                    2937
customer_lifetime_value      2940
income                       2937
monthly_premium_auto         2937
number_of_open_complaints    2937
policy_type                  2937
vehicle_class                2937
total_claim_amount           2937
dtype: int64


In [6]:
#drop all the lines with nulls
df = df.dropna(how = 'all')
null_counts_after_drop = df.isnull().sum()
print(null_counts_after_drop)

customer                       0
state                          0
gender                       117
education                      0
customer_lifetime_value        3
income                         0
monthly_premium_auto           0
number_of_open_complaints      0
policy_type                    0
vehicle_class                  0
total_claim_amount             0
dtype: int64


In [7]:
# Display rows where 'customer_lifetime_value' has null values
rows_with_nulls = df[df['customer_lifetime_value'].isnull()]

# Print the resulting dataframe
display(rows_with_nulls)

customer       state gender             education  \
0    RB50392  Washington    NaN                Master   
78   SP81997  Washington      F                Master   
988  GT62080  Washington      M  High School or Below   

     customer_lifetime_value   income  monthly_premium_auto  \
0                        NaN      0.0                1000.0   
78                       NaN  41275.0                  96.0   
988                      NaN  55561.0                  63.0   

     number_of_open_complaints    policy_type  vehicle_class  \
0                          0.0  Personal Auto  Four-Door Car   
78                         0.0  Personal Auto  Four-Door Car   
988                        0.0  Personal Auto  Four-Door Car   

     total_claim_amount  
0              2.704934  
78            41.122303  
988          227.872071

In [8]:
#Remove the row with index 0
df = df.drop(0)

# Print the resulting dataframe
display(rows_with_nulls)

customer       state gender             education  \
0    RB50392  Washington    NaN                Master   
78   SP81997  Washington      F                Master   
988  GT62080  Washington      M  High School or Below   

     customer_lifetime_value   income  monthly_premium_auto  \
0                        NaN      0.0                1000.0   
78                       NaN  41275.0                  96.0   
988                      NaN  55561.0                  63.0   

     number_of_open_complaints    policy_type  vehicle_class  \
0                          0.0  Personal Auto  Four-Door Car   
78                         0.0  Personal Auto  Four-Door Car   
988                        0.0  Personal Auto  Four-Door Car   

     total_claim_amount  
0              2.704934  
78            41.122303  
988          227.872071

In [9]:
# Fill null values in customer_lifetime_value with the mean
mean_value = df['customer_lifetime_value'].mean()
df['customer_lifetime_value'].fillna(mean_value, inplace=True)
rows_with_nulls = df[df['customer_lifetime_value'].isnull()]
display(rows_with_nulls)

Empty DataFrame
Columns: [customer, state, gender, education, customer_lifetime_value, income, monthly_premium_auto, number_of_open_complaints, policy_type, vehicle_class, total_claim_amount]
Index: []

In [10]:
# Fill null values in 'gender' with 'U'
df['gender'].fillna('U', inplace=True)
null_counts_after_drop = df.isnull().sum()
display(null_counts_after_drop)

customer                     0
state                        0
gender                       0
education                    0
customer_lifetime_value      0
income                       0
monthly_premium_auto         0
number_of_open_complaints    0
policy_type                  0
vehicle_class                0
total_claim_amount           0
dtype: int64

After identifying the null values, we can see that all the columns have at least 2937 nulls. Therefore it's better to drop all the lines with nulls, instead of filling them with any value. With this cleaning measure we are left with 117 nulls in 'gender' and 3 nulls in 'customer_lifetime_value'.

After closer look at the nulls in 'customer_lifetime_value' we can eleminate the row [0], as it has multiple null values in other columns. 

As I didn't provide nessesary measures to check, if the data has a skewed distribution or outliers, I decided to fill the two rest null values with mean value. 

The null values in 'gender' were replaces with 'U'(unknown)

## Exercise 5: Dealing with duplicates

In [11]:
# Check for duplicate rows
duplicate_rows = df[df.duplicated()]
duplicate_customer = df[df['customer'].duplicated()]

# Display the duplicate rows, if any
print("Duplicate Rows:")
display(duplicate_rows)
display(duplicate_customer)


Duplicate Rows:


Empty DataFrame
Columns: [customer, state, gender, education, customer_lifetime_value, income, monthly_premium_auto, number_of_open_complaints, policy_type, vehicle_class, total_claim_amount]
Index: []

Empty DataFrame
Columns: [customer, state, gender, education, customer_lifetime_value, income, monthly_premium_auto, number_of_open_complaints, policy_type, vehicle_class, total_claim_amount]
Index: []

After checking the data set for duplicates, both the whole row and the column 'Customer' separatelly, no duplicates were found.

# Bonus: Challenge 2: creating functions on a separate `py` file

In [ ]:
#Functions.py

# Bonus: Challenge 3: Analyzing Clean and Formated Data

In [18]:
# Review statistics for total claim amount and customer lifetime value
total_claim_stats = df[['total_claim_amount']].describe()
customer_lifetime_stats = df[['customer_lifetime_value']].describe()

# Identify top 25% of total claim amount and bottom 25% of customer lifetime value
high_claim_threshold = df['total_claim_amount'].quantile(0.75)
low_clv_threshold = df['customer_lifetime_value'].quantile(0.25)

# Create DataFrame with high policy claim amount and low customer lifetime value
high_claim_low_clv_df = df[(df['total_claim_amount'] > high_claim_threshold) & 
                           (df['customer_lifetime_value'] < low_clv_threshold)]

# Display summary statistics for the selected data
selected_data_stats = high_claim_low_clv_df[['total_claim_amount', 'customer_lifetime_value']].describe()

# Display results
print("Summary Statistics for Total Claim Amount:")
display(total_claim_stats)

print("\nSummary Statistics for Customer Lifetime Value:")
display(customer_lifetime_stats)

print("\nSummary Statistics for High Policy Claim Amount and Low Customer Lifetime Value:")
display(selected_data_stats)

Summary Statistics for Total Claim Amount:


total_claim_amount
count         1070.000000
mean           405.362874
std            292.905738
min              0.382107
25%            203.264086
50%            354.964564
75%            532.800000
max           2893.239678


Summary Statistics for Customer Lifetime Value:


customer_lifetime_value
count             1.070000e+03
mean              7.936903e+05
std               6.428762e+05
min               2.004351e+05
25%               4.043018e+05
50%               5.885745e+05
75%               8.952745e+05
max               5.816655e+06


Summary Statistics for High Policy Claim Amount and Low Customer Lifetime Value:


total_claim_amount  customer_lifetime_value
count           40.000000                40.000000
mean           723.776577            341577.910000
std            149.280606             56921.472027
min            537.600000            228759.690000
25%            633.378846            304220.875000
50%            685.284189            365995.370000
75%            799.200000            387364.700000
max           1185.988301            402381.440000

The mean total claim amount for all customers is 405.36, while for the subset with high policy claim amount and low customer lifetime value, the mean is 723.78. This shows that customers in the subset on average, tend to file claims that are significantly larger compared to the average customer. This could imply that these customers may have more complex or costly claims.
The mean customer lifetime value for all customers is 793,690.3, whereas for the subset, the mean is 341,577.91.